### ISSUE TITLE
Positivity check should not ignore all effects

### ISSUE THREAD

**Issue by mtzguido** [FStarLang/FStar#1141]
Currently, the positivity check for inductives is disabled (assumed valid for a given constructor) when the negativity occurs with a non-Tot (nor GTot) effect in its codomain, thus, types such as
```
noeq type negative =
  | Blah : (negative -> Dv negative) -> negative
```
are accepted, which is pretty sweet since they would never compromise consistency anyway, as a `DIV` computation can never become `Tot`. However, given our DM4F magic, some effects can actually disappear into `Tot`, and we should check positivity of them. Here's a proof of False exploiting that:

```
module Bug

let id (a:Type) = unit -> M a

val return_id : (a:Type) -> (x:a) -> id a
let return_id a x = fun _ -> x

val bind_id : (a:Type) -> (b:Type) -> (f:id a) -> (g:a -> id b) -> id b
let bind_id a b f g = fun _ ->
  let r = f () in g r ()

total reifiable reflectable new_effect {
  IDN : (a:Type) -> Effect
  with repr     = id
     ; bind     = bind_id
     ; return   = return_id
}

effect ID (a:Type) = IDN a (fun () p -> forall x. p x)

unopteq
type np (a : Type) =
    | Mk : (np a -> ID a) -> np a

val self : (#a:Type) -> np a -> ID a
let self #a v =
    match v with
    | Mk f -> f v

val loop : (#a:Type) -> (a -> a) -> a
let loop #a f = reify (self (Mk (fun x -> f (self x)))) ()

val falso : squash False
let falso = loop (fun x -> x)
```
(this is also a diverging Tot term)

In short, maybe we should really check if the effect we're traversing is marked as total or not?  This proof works the same without the `total` flag in the dm4f declaration, but that could be fixed by reifying non-total DM4F effects into Div instead of Tot, right?

**Comment by mtzguido** [comment]
Making a note that this is more pervasive than it seems, `TOTAL` flags are being checked but we seem to not be keeping/generating them correctly. The following works without even using DM4F:
```
effect MyTot (a:Type) = PURE a (fun p -> forall x. p x)

noeq
type np (a : Type) = | Mk : (np a -> MyTot a) -> np a

let self #a (v : np a) : a =
    match v with
    | Mk f -> f v

let loop #a (f : a -> a) : a =
    self (Mk (fun x -> f (self x)))

val falso : squash False
let falso = loop (fun x -> x)
```
Will take a deeper look later.

**Comment by kyoDralliam** [comment]
We shouldn't check for `TOTAL` but anything above `PARTIAL_RETURN` no ? From what I understand `TOTAL` is given to computation types which are syntactically derived from `Tot` or `GTot`, never for computations coming from `Pure` (I don't think we are checking that the wp is always satisfiable)

**Comment by mtzguido** [comment]
I wasn't aware of the difference to be honest :), but yeah, whichever one means "terminating". From what I could see both `ID` and `MyTot` end up with an empty set of flags (meaning they're taken as diverging). Should they be marked as `PARTIAL_RETURN`?

I agree that flipping the logic could make more sense, I find it a bit confusing currently.

**Comment by nikswamy** [comment]
@aseemr : the positivity check is in your territory. It seems a relatively easy, though very important fix, to apply.

**Comment by aseemr** [comment]
Thanks for the reports @mtzguido. This is fixed in master, both the examples fail now.

**Comment by mtzguido** [comment]
Just looked at this again with @kyoDralliam and seems we missed a spot: we're still reifying non-`total` effects to `Tot`, which is unsound in two ways:

1. Just recursively prove false in the effect and reify it
```fstar
module Bug1141d

let id (a:Type) = unit -> M a

val return_id : (a:Type) -> (x:a) -> id a
let return_id a x = fun _ -> x

val bind_id : (a:Type) -> (b:Type) -> (f:id a) -> (g:a -> id b) -> id b
let bind_id a b f g = fun _ ->
  let r = f () in g r ()

// NB: no `total` keyword, so no positivity check
reifiable reflectable new_effect {
  IDN : (a:Type) -> Effect
  with repr     = id
     ; bind     = bind_id
     ; return   = return_id
}

effect ID (a:Type) = IDN a (fun () p -> forall x. p x)

let rec f () : ID False = f ()

let x : False = reify (f ()) ()
```

2. Bypass the termination check, and do as above:
```fstar
module Bug1141c

let id (a:Type) = unit -> M a

val return_id : (a:Type) -> (x:a) -> id a
let return_id a x = fun _ -> x

val bind_id : (a:Type) -> (b:Type) -> (f:id a) -> (g:a -> id b) -> id b
let bind_id a b f g = fun _ ->
  let r = f () in g r ()

// NB: no `total` keyword, so no positivity check
reifiable reflectable new_effect {
  IDN : (a:Type) -> Effect
  with repr     = id
     ; bind     = bind_id
     ; return   = return_id
}

effect ID (a:Type) = IDN a (fun () p -> forall x. p x)

noeq
type np (a : Type) =
    | Mk : (np a -> ID a) -> np a

val self : (#a:Type) -> np a -> ID a
let self #a v =
    match v with
    | Mk f -> f v

val loop : (#a:Type) -> (a -> a) -> a
let loop #a f = reify (self (Mk (fun x -> f (self x)))) ()

val falso : squash False
let falso = loop (fun x -> x)
```

**Comment by mtzguido** [comment]
I have a fix for this, waiting for CI. The idea is that for a non-total DM4F effect `E`, with `e : E t wp`, then `reify e : Dv (reprE t wp)`. I was previously thinking of changing repr to use `DIV` instead of `PURE` (so, for state, we'd get `repr t wp = s:st -> DIV (t * s) (fun ts1 -> wp s0 ts1)`, but that's not enough; consider:

```fstar
let rec f () : ST unit _ = f ()

let x = reify (f ())
```
The definition of `x` is already looping (e.g. after extracting), so it should be marked as divergent. In this case one can eta expand, but not in general. Or so I think, at least, since we have products on C types. Not sure it's useful to have that as a top-level (and it didn't really work when I tried to do it).

Note that this implies any non-`total` effect cannot be reified to prove lemmas about it. Doesn't seem to be an issue for code in CI (I added some `total` flags and that was it).

**Comment by aseemr** [comment]
Just looking at it, thanks Guido! This makes sense.

Though I don't see this as a bug in positivity checking, rather in the reification of non-total dm4f effects to `Tot`. As far as I can see both the repros and the fix itself seem to be related to that.

**Comment by mtzguido** [comment]
Yes :) maybe a separate issue was more appropriate, sorry.
